src/liballoc/arc.rs

```diff
@@ -54,16 +54,32 @@ const MAX_REFCOUNT: usize = (isize::MAX) as usize;
 /// exception. If you need to mutate through an `Arc`, use [`Mutex`][mutex],
 /// [`RwLock`][rwlock], or one of the [`Atomic`][atomic] types.
 ///
-/// `Arc` uses atomic operations for reference counting, so `Arc`s can be
-/// sent between threads. In other words, `Arc<T>` implements [`Send`]
-/// as long as `T` implements [`Send`] and [`Sync`][sync]. The disadvantage is
-/// that atomic operations are more expensive than ordinary memory accesses.
-/// If you are not sharing reference-counted values between threads, consider
-/// using [`rc::Rc`][`Rc`] for lower overhead. [`Rc`] is a safe default, because
-/// the compiler will catch any attempt to send an [`Rc`] between threads.
-/// However, a library might choose `Arc` in order to give library consumers
+/// ## Thread Safety
+///
+/// Unlike [`Rc<T>`], `Arc<T>` uses atomic operations for its reference
+/// counting.  This means that it is thread-safe. The disadvantage is that
+/// atomic operations are more expensive than ordinary memory accesses.  If you
+/// are not sharing reference-counted values between threads, consider using
+/// [`Rc<T>`] for lower overhead. [`Rc<T>`] is a safe default, because the
+/// compiler will catch any attempt to send an [`Rc<T>`] between threads.
+/// However, a library might choose `Arc<T>` in order to give library consumers
 /// more flexibility.
 ///
+/// `Arc<T>` will implement [`Send`] and [`Sync`] as long as the `T` implements
+/// [`Send`] and [`Sync`]. This may be a bit counter-intuitive at first: after
+/// all, isn't the point of `Arc<T>` thread safety? The key is this: `Arc<T>`
+/// is an implementation of multiple ownership that is thread safe, but it
```

Comment:
If you had a type `T` that is `!Sync` and an `Arc<T>` that does not give you access to the internal object, then the `Arc<T>` can be `Send`, because you can never access the inner data, but you can clone and drop the `Arc<T>` to your heart's desire. This is the guarantee that `Arc<T>` gives, without looking at the `T`, you can clone and drop the `Arc<T>`. Since `Arc<T>` also gives you access to it's inner data through `Deref`, you also need the `Sync` bound on `T` to make `Arc<T>` `Send`, but that's nothing special about `Arc<T>` but a general thing about sharing data over thread boundaries as stated in the `Sync` docs.